osquery/events/windows/etw/etw_kernel_session.cpp

```diff
@@ -0,0 +1,212 @@
+  } break;
+
+  default:
+    return Status::failure("Unsupported kernel provider was provided.");
+  }
+
+  if (kernelProvider == nullptr) {
+    return Status::failure("There was a problem with given kernel provider.");
+  }
+
+  // Provider preprocessor callback is registered
+  kernelProvider->add_on_event_callback(configData.getPreProcessor());
+
+  // Pausing the trace event session to enable the new provider configuration
+  pause();
+  runningProviders_.push_back(kernelProvider);
+  kernelTraceSession_->enable(*kernelProvider);
+
+  // Resume the trace session listening once provider is ready
+  resume();
+
+  return Status::success();
+}
+
+void KernelEtwSessionRunnable::start() {
+  std::unique_lock<std::mutex> lock(mutex_);
+  if (kernelTraceSession_) {
+    while (!endTraceSession_) {
+      kernelTraceSession_->start();
+      traceSessionStopped_ = true;
+
+      if (!endTraceSession_) {
+        condition_.wait(lock);
+      }
+    }
+  }
+}
+
+void KernelEtwSessionRunnable::stop() {
+  if (kernelTraceSession_) {
+    endTraceSession_ = true;
+    kernelTraceSession_->stop();
+  }
+}
+
+void KernelEtwSessionRunnable::pause() {
+  if (kernelTraceSession_) {
+    kernelTraceSession_->stop();
+    while (!traceSessionStopped_) {
+      Sleep(500);
+    }
+    traceSessionStopped_ = false;
+  }
+}
+
+void KernelEtwSessionRunnable::resume() {
+  if (kernelTraceSession_) {
+    condition_.notify_one();
+  }
+}
+
+void KernelEtwSessionRunnable::initKernelTraceSession(
+    const std::string& sessionName) {
+  if (sessionName.empty()) {
+    return;
+  }
+
+  // stop trace session
+  stopKernelTraceSession(sessionName.c_str());
+
+  // check if this can fail
+  kernelTraceSession_ =
+      std::make_shared<krabs::kernel_trace>(stringToWstring(sessionName));
+
+  // Setting default trace session properties
+  EVENT_TRACE_PROPERTIES session_properties = {0};
+  session_properties.BufferSize = FLAGS_etw_kernel_trace_buffer_size;
+  session_properties.MinimumBuffers = FLAGS_etw_kernel_trace_minimum_buffers;
+  session_properties.MaximumBuffers = FLAGS_etw_kernel_trace_maximum_buffers;
+  session_properties.FlushTimer = FLAGS_etw_kernel_trace_flush_timer;
+  session_properties.LogFileMode = kEventTraceLogFileMode;
+  kernelTraceSession_->set_trace_properties(&session_properties);
+}
+
+void KernelEtwSessionRunnable::stopKernelTraceSession(
+    const std::string& sessionName) {
+  if (sessionName.empty()) {
+    return;
+  }
+
+  struct SessionData {
+    EVENT_TRACE_PROPERTIES Properties;
+    WCHAR SessionName[MAX_PATH];
+  };
+
+  SessionData sessionInfo = {0};
+  sessionInfo.Properties.Wnode.BufferSize =
+      (ULONG)sizeof(EVENT_TRACE_PROPERTIES);
+  sessionInfo.Properties.Wnode.Flags = WNODE_FLAG_TRACED_GUID;
+  sessionInfo.Properties.LoggerNameOffset =
+      (ULONG)sizeof(EVENT_TRACE_PROPERTIES);
+
+  /// Best effort to stop ongoing trace session
+  ControlTraceA(NULL,
```

Comment:
good idea, this is inside of a best-effort helper function, so I'm just capturing the return code of ControlTrace() and logging it in case the call failed